File: Python/linkList.py

class ListNode:
    def __init__(self, data: int | None = None, next: 'ListNode | None' = None):
        self.data = data
        self.next = next

class LinkList:
    def __init__(self):
        self.head = ListNode()

    def add(self, value: int):
        new_node = ListNode(value)
        node_current = self.head
        while (node_current.next):
            node_current = node_current.next
        node_current.next = new_node

    def out(self):
        node_current = self.head
        while node_current:
            print(f"{node_current.data}", end=" -> ")
            node_current = node_current.next
        print()




def main():
    linkList = LinkList()
    linkList.add(1)
    linkList.add(2)
    linkList.add(3)
    linkList.add(4)
    linkList.add(5)
    linkList.out()

if __name__ == "__main__":
    main()

File: c/linkList.c
#include <stdio.h>
#include <limits.h>
#include <stdlib.h>


typedef struct Node 
{
	int data;
	struct Node *next;
} Node;

Node *createNewNode(int value)
{
	Node *new_node = malloc(sizeof(Node));
	if (new_node == NULL)
	{
		fprintf(stderr, "Allocation memory for new node failed\n");
		exit(1);
	}
	new_node->data = value;
	new_node->next = NULL;
	return new_node;


}

void addNewNodeToTail(Node **tail, int value)
{
	Node *new_node = createNewNode(value);
	(*tail)->next = new_node;
	*tail = (*tail)->next;
}


Node *initLinkList()
{
	Node *head = malloc(sizeof(Node));
	if (head == NULL)
	{
		fprintf(stderr, "Allocation memory for head failed\n");
		exit(1);
	}
	head->data = INT_MIN;
	head->next = NULL;
	return head;
}


void out(Node *head)
{
	Node *ptr = head;
	while (ptr != NULL)
	{
		printf("%d", ptr->data);
		if (ptr->next != NULL) printf(" -> ");
		ptr = ptr->next;
	}
	printf("\n");
}

void freeLinkList(Node *head)
{
	Node *ptr = head;
	while (ptr != NULL)
	{
		Node *tmp = ptr;
		ptr = ptr->next;
		free(tmp);
	}
}

int main(void)
{
	Node *head = initLinkList();
	Node *tail = head;

	addNewNodeToTail(&tail, 1);
	addNewNodeToTail(&tail, 2);
	addNewNodeToTail(&tail, 3);
	addNewNodeToTail(&tail, 4);

	out(head);
	freeLinkList(head);
	return 0;
}
